keps/77-dynamically-sized-jobs/README.md

```diff
@@ -0,0 +1,255 @@
+
+For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.
+
+### User Stories
+
+#### Story 1 - RayCluster w/ autoscaling
+
+1. The user creates a RayCluster with `enableInTreeAutoscaling: true`.
+2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 
+3. Given the deman of the job, the Ray Autoscaler adjusts the replicas field as it adds or removes Pods from the cluster.
+4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 
+
+### Notes/Constraints/Caveats (Optional)
+
+If Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.
+
+## Design Details
+
+### Workload Slices
+To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.
+
+The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.
+
+
+### Creating Workload Slices
+
+The [GenericJob interface](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go#L30-L55) will be updated to handle resize operations of jobs.
+
+```golang
+type GenericJob interface {
+  ...
+  ...
+  ResizeJob(wl *kueue.Workload) error
+}
+```
+Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 
+
+### Pod Scheduling Gates
+
+Inside **raycluster_webhook** implement schedulingGate injection for pods on RayCluster creation time.
+The Pods will be ungated following a similar behavior as to how a job is suspended and then unsuspended in the when admitted.
+When the RayCluster scales up, the new pods will be gated due to the schedulingGates injection in the webhook.
+
+After the creation of each individual Workload Slice and admission of a Workload Slice, the **workload_scheduling_gates_controller** should be in charge of removing the scheduling gates from each pod. We only need to ungate the number of pods to match the number of admitted pods, this should be a counter. We don’t want to accidentally ungate too many pods since race conditions could happen and we also don’t want to double count. It's worth mentioning that for the case of recreated pods (i.e. machine failure for example), these pods will go through the admission/scheduuling check again, Kueue is responsible fo removing the scheduling gates when there's available quota and resources to spend on the RayCluster. 
```

Comment:
Are all pods created by RayCluster identical?  If not, how will we tell recreated pods that should be admitted immediately from "additional" pods that should wait?

Reply:
> Are all pods created by RayCluster identical?

All worker pods from the same worker group share the same pod template, so it doesn't really matter if newly created pods or older pods are admitted.

Reply:
Cool :). Please add this information to the KEP (if it is not there already).

Reply:
Added